github/messages.go

```diff
@@ -132,17 +138,40 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
 //       // Process payload...
 //     }
 //
-func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	switch c := r.Header.Get("Content-Type"); c {
+	case "application/x-www-form-urlencoded":
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
+	case "application/json":
+		defer r.Body.Close()
+		var err error
+		jsonPayload, err = ioutil.ReadAll(r.Body)
+		// If the content type is application/json then the raw body
+		// used to calculate the signature is just the original body.
+		rawPayload = jsonPayload
+		if err != nil {
```

Comment:
This `if err != nil` check is out of place. Did you mean to place it directly under `jsonPayload, err = ioutil.ReadAll(r.Body)` line?